libs/mimir/src/adapters/primary/bragi/handlers.rs

```diff
@@ -27,29 +31,90 @@ use places::{addr::Addr, admin::Admin, poi::Poi, stop::Stop, street::Street, Pla
     params: ForwardGeocoderQuery,
     geometry: Option<Geometry>,
-    client: S,
+    client: ElasticsearchStorage,
     settings: settings::QuerySettings,
-) -> Result<impl warp::Reply, warp::Rejection>
-where
-    S: SearchDocuments,
-    S::Document: Serialize + Into<serde_json::Value>,
-{
+) -> Result<impl warp::Reply, warp::Rejection> {
     let q = params.q.clone();
     let timeout = params.timeout;
     let es_indices_to_search_in =
         build_es_indices_to_search(&params.types, &params.pt_dataset, &params.poi_dataset);
     let lang = params.lang.clone().unwrap_or_else(|| "fr".to_string());
     let filters = filters::Filters::from((params, geometry));
-    let dsl = dsl::build_query(&q, filters.clone(), lang, &settings);
+    let dsl_query_prefix = dsl::build_query(
+        &q,
+        filters.clone(),
+        lang.clone(),
+        &settings,
+        QueryType::PREFIX,
+    );
+    let result_prefix_query = send_query(
+        client.clone(),
+        dsl_query_prefix,
+        timeout,
+        es_indices_to_search_in.clone(),
+        filters.clone(),
+        &q,
+    )
+    .await;
 
+    match result_prefix_query {
+        Ok(resp) => {
+            if resp.features.is_empty() {
+                let dsl_query_fuzzy = dsl::build_query(
+                    &q,
+                    filters.clone(),
+                    lang.clone(),
+                    &settings,
+                    QueryType::FUZZY,
+                );
+                let result_fuzzy_query = send_query(
+                    client,
+                    dsl_query_fuzzy,
+                    timeout,
+                    es_indices_to_search_in,
+                    filters,
+                    &q,
+                )
+                .await;
+                match result_fuzzy_query {
+                    Ok(resp) => Ok(with_status(json(&resp), StatusCode::OK)),
+                    Err(err) => Ok(with_status(
+                        json(&format!(
+                            "Error while searching {}: {}",
+                            &q,
+                            err.to_string()
+                        )),
+                        err,
+                    )),
+                }
+            } else {
+                println!("Prefix query");
+                Ok(with_status(json(&resp), StatusCode::OK))
+            }
+        }
+        Err(_err) => Ok(with_status(
+            json(&format!("Error while searching {}", &q)),
+            StatusCode::INTERNAL_SERVER_ERROR,
+        )),
+    }
+}
+
+async fn send_query(
+    client: ElasticsearchStorage,
+    dsl: serde_json::Value,
+    timeout: Option<Duration>,
+    es_indices_to_search_in: Vec<String>,
+    filters: Filters,
+    q: &str,
+) -> Result<GeocodeJsonResponse, StatusCode> {
     debug!("{}", serde_json::to_string(&dsl).unwrap());
 
     match client
         .search_documents(
-            es_indices_to_search_in,
+            es_indices_to_search_in.clone(),
```

Comment:
```suggestion
            es_indices_to_search_in,
```
since you don't use `es_indices_to_search_in` afterwards